tests/attrs_test.py

```diff
@@ -208,5 +209,145 @@ def body(x, __):
     self.assertAllClose(thing.x, 1024., check_dtypes=False)
 
 
+class AttrsJVPTest(jtu.JaxTestCase):
+
+  @parameterized.parameters([True, False])
+  def test_jvp_basic(self, jit):
+    thing = Thing(2.0)
+
+    def f():
+      x = jax_getattr(thing, 'x')
+      x = jnp.sin(x)
+      jax_setattr(thing, 'x', x)
+
+    if jit:
+      f = jax.jit(f)
+
+    _, _, attr_tangents = attrs.jvp(f, (), (), [(thing, 'x', 1.0)])
+    self.assertAllClose(thing.x, jnp.sin(2.0), check_dtypes=False)
+    (thing_, attr_, tangent_), = attr_tangents
+    self.assertIs(thing, thing_)
+    self.assertEqual(attr_, 'x')
+    self.assertAllClose(tangent_, jnp.cos(2.0), check_dtypes=False)
+
+  @parameterized.parameters([True, False])
+  def test_jvp_clobber(self, jit):
+    thing = Thing(2.0)
+
+    def f():
+      x = jax_getattr(thing, 'x')
+      x = jnp.sin(2.0)
+      jax_setattr(thing, 'x', x)
+
+    if jit:
+      f = jax.jit(f)
+
+    _, _, attr_tangents = attrs.jvp(f, (), (), [(thing, 'x', 1.0)])
+    self.assertAllClose(thing.x, jnp.sin(2.0), check_dtypes=False)
+    self.assertEmpty(attr_tangents)
+
+  @parameterized.parameters([True, False])
+  def test_jvp_nowrite(self, jit):
+    thing = Thing(2.0)
+
+    def f():
+      x = jax_getattr(thing, 'x')
+
+    if jit:
+      f = jax.jit(f)
+
+    _, _, attr_tangents = attrs.jvp(f, (), (), [(thing, 'x', 1.0)])
+    self.assertAllClose(thing.x, 2.0, check_dtypes=False)
+    (thing_, attr_, tangent_), = attr_tangents
+    self.assertIs(thing, thing_)
+    self.assertEqual(attr_, 'x')
+    self.assertAllClose(tangent_, 1.0, check_dtypes=False)
+
+  def test_jit_of_jvp(self):
+    thing = Thing(2.0)
+
+    def f():
+      x = jax_getattr(thing, 'x')
+      x = jnp.sin(x)
+      jax_setattr(thing, 'x', x)
+
+    @jax.jit
+    def g():
+      _, _, attr_tangents = attrs.jvp(f, (), (), [(thing, 'x', 1.0)])
+      (thing_, attr_, tangent_), = attr_tangents
+      self.assertIs(thing, thing_)
+      self.assertEqual(attr_, 'x')
+      return jax_getattr(thing, 'x'), tangent_
+
+    x, tangent = g()
+    self.assertAllClose(x, jnp.sin(2.0), check_dtypes=False)
+    self.assertAllClose(tangent, jnp.cos(2.0), check_dtypes=False)
+
+  @parameterized.parameters([True, False])
+  def test_jvp_higher_order(self, jit):
```

Comment:
I should've written a general `check_grads` function but I had to get a feel for the API by doing this manually first. I'll write one at some point!